File: 2025-2029/Adithyanandan/openagent-hub/backend/app/api/tokens.py
from uuid import UUID

from fastapi import APIRouter, Depends, HTTPException
from fastapi.security import HTTPBearer, HTTPAuthorizationCredentials
from sqlalchemy.orm import Session

from app.core.database import get_db
from app.schemas.api_token import ApiTokenCreate, ApiTokenResponse, ApiTokenCreated
from app.services.auth_service import get_current_user
from app.services.api_token_service import (
    list_tokens,
    create_token,
    revoke_token,
)

router = APIRouter(prefix="/tokens", tags=["tokens"])
security = HTTPBearer()


def _current_user(
    credentials: HTTPAuthorizationCredentials = Depends(security),
    db: Session = Depends(get_db),
):
    return get_current_user(db, credentials.credentials)


@router.get("", response_model=list[ApiTokenResponse])
def get_tokens(user=Depends(_current_user), db: Session = Depends(get_db)):
    return list_tokens(db, user.id)


@router.post("", response_model=ApiTokenCreated, status_code=201)
def add_token(data: ApiTokenCreate, user=Depends(_current_user), db: Session = Depends(get_db)):
    row, plaintext = create_token(db, user.id, data.name)
    return ApiTokenCreated(
        id=row.id,
        name=row.name,
        prefix=row.prefix,
        revoked=row.revoked,
        last_used_at=row.last_used_at,
        created_at=row.created_at,
        token=plaintext,
    )


@router.delete("/{token_id}", status_code=204)
def remove_token(token_id: UUID, user=Depends(_current_user), db: Session = Depends(get_db)):
    try:
        revoke_token(db, user.id, token_id)
    except ValueError as e:
        raise HTTPException(status_code=404, detail=str(e))

File: 2025-2029/Adithyanandan/openagent-hub/frontend/src/services/tokens.ts
import api from './api';

export interface ApiToken {
  id: string;
  name: string;
  prefix: string;
  revoked: boolean;
  last_used_at: string | null;
  created_at: string | null;
}

export interface ApiTokenCreated extends ApiToken {
  token: string; // plaintext, shown only once
}

export async function listTokens(): Promise<ApiToken[]> {
  const { data } = await api.get('/tokens');
  return data;
}

export async function createToken(name: string): Promise<ApiTokenCreated> {
  const { data } = await api.post('/tokens', { name });
  return data;
}

export async function revokeToken(id: string): Promise<void> {
  await api.delete(`/tokens/${id}`);
}
